# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [74]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline
from sklearn import datasets
os.getcwd()

'C:\\Users\\chris\\Springboard_local\\6_ApplyDSM\\6_4_PreProcess_TrainingDevelop\\data'

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [75]:
path=r"C:\Users\chris\Springboard_local\6_ApplyDSM\6_4_PreProcess_TrainingDevelop\data"
os.chdir(path) 
os.getcwd()
df = pd.read_csv('step3_output.csv')
df.head()
df.shape

(330, 30)

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [76]:
#df2 = df.drop(['Name', 'Region', 'state', 'chair_sum', 'sum_check'], axis=1)
df = pd.concat([df.drop(['Region','state'], axis=1), pd.get_dummies(df['state'])], axis=1)
df.head()
#print(df2.head())


Name  summit_elev  vertical_drop  base_elev  trams  \
0           Aspen / Snowmass        12510           4406       8104      3   
1         Deer Valley Resort         9570           3000       6570      1   
2     Copper Mountain Resort        12313           2738       9712      1   
3  Mammoth Mountain Ski Area        11053           3100       7953      3   
4                  Telluride        13150           4425       8725      2   

   fastEight  fastSixes  fastQuads  quad  triple  ...  Rhode Island  \
0        0.0          1         15     4       3  ...             0   
1        0.0          0         13     0       5  ...             0   
2        0.0          2          4     0       4  ...             0   
3        0.0          2          9     1       6  ...             0   
4        0.0          0          6     1       2  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     1        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  

[5 rows x 63 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [77]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend', 'chair_sum', 'sum_check'], axis=1)
# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df['AdultWeekend']
# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [78]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()
# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [79]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [80]:
from sklearn.ensemble import RandomForestRegressor
regressor = RandomForestRegressor(random_state=0, n_estimators=200)
regressor.fit(X_train, y_train)
regressor.score(X_test, y_test)

0.8778766643388571

In [81]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = regressor.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [82]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [83]:
from sklearn.metrics import explained_variance_score, mean_absolute_error
print(explained_variance_score(y_test, y_pred))
print(mean_absolute_error(y_test, y_pred))

0.8797496026970408
6.270051075268832


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [84]:
print(lm.intercept_)

155324525245.58307


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [85]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
dfco = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
dfco.sort_values(by='Coefficient', ascending=False, inplace=True)
dfco.head(15)


Coefficient
total_chairs   3.179247e+13
fastQuads      1.205260e+13
surface        1.129238e+13
double         9.951264e+12
triple         8.877211e+12
quad           7.194651e+12
New York       5.125709e+12
Michigan       4.837278e+12
Colorado       4.261922e+12
California     4.170688e+12
Pennsylvania   3.979933e+12
Wisconsin      3.669810e+12
New Hampshire  3.669810e+12
fastSixes      3.572999e+12
Vermont        3.558932e+12

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [86]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

X = df2
# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df2['AdultWeekend']
# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()
# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)
from sklearn.metrics import explained_variance_score, mean_absolute_error

In [87]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
from sklearn.ensemble import RandomForestRegressor
regressor = RandomForestRegressor(random_state=0, n_estimators=200)
regressor.fit(X_train, y_train)
regressor.score(X_test, y_test)
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = regressor.predict(X_test)
print(lm.intercept_)
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
dfco = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
dfco.sort_values(by='Coefficient', ascending=False, inplace=True)
from sklearn.metrics import explained_variance_score, mean_absolute_error
print(explained_variance_score(y_test, y_pred))
print(mean_absolute_error(y_test, y_pred))
dfco.head(15)

64.16681003584227
0.9711272230992706
1.2938331562810559


Coefficient
AdultWeekend     2.253710e+01
surface          3.012333e-14
clusters         2.985244e-14
total_chairs     2.743196e-14
summit_elev      2.489660e-14
AdultWeekday     2.465821e-14
TerrainParks     2.084317e-14
base_elev        1.685320e-14
averageSnowfall  1.632653e-14
fastEight        1.631328e-14
NightSkiing_ac   1.337451e-14
fastSixes        1.237318e-14
Snow Making_ac   1.140782e-14
trams            9.625542e-15
quad             9.459569e-15

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [88]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

df3 = df2.drop(['summit_elev', 'base_elev'], axis=1)
X = df3
# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df3['AdultWeekend']
# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()
# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)
from sklearn.metrics import explained_variance_score, mean_absolute_error

In [89]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
from sklearn.ensemble import RandomForestRegressor
regressor = RandomForestRegressor(random_state=0, n_estimators=200)
regressor.fit(X_train, y_train)
regressor.score(X_test, y_test)
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = regressor.predict(X_test)
print(lm.intercept_)
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
dfco = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
dfco.sort_values(by='Coefficient', ascending=False, inplace=True)
from sklearn.metrics import explained_variance_score, mean_absolute_error
print(explained_variance_score(y_test, y_pred))
print(mean_absolute_error(y_test, y_pred))
dfco.head(15)

64.16681003584229
0.971756587610182
1.2403814893984666


Coefficient
AdultWeekend       2.253710e+01
yearsOpen          2.318520e-14
daysOpenLastYear   1.383595e-14
projectedDaysOpen  1.131967e-14
Runs               1.065640e-14
surface            1.053251e-14
vertical_drop      9.424407e-15
AdultWeekday       9.189364e-15
double             9.133266e-15
Snow Making_ac     7.714868e-15
total_chairs       7.268384e-15
trams              6.673177e-15
fastSixes          6.232166e-15
clusters           4.305451e-15
quad               3.433440e-15

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.8797 |6.27  |-|
| Model 2. | 0.9711|1.2938 |'state'|
| Model 3. | 0.9717 |1.2403 |'state','summit_elev','base_elev'|

Model Selection:

In [ ]:
#Model 3 would be the best selection based on explained variance the mean absolute error.